### PR TITLE
[SPARK-38650][SQL] Better ParseException message for char types without length

### PR DESCRIPTION
**Pull request by anchovYu**
### What changes were proposed in this pull request?

This PR improves the error messages for the char / varchar / character datatypes without length. It also added related testcases.

#### Details
We support char and varchar types. But when users input the type without length, the message is confusing and not helpful at all:

```
> SELECT cast('a' as CHAR)

DataType char is not supported.(line 1, pos 19)

== SQL ==
SELECT cast('a' AS CHAR)
-------------------^^^
```
In the after case, the messages would be:
```
Datatype char requires a length parameter, for example char(10). Please specify the length.

== SQL ==
SELECT cast('a' AS CHAR)
-------------------^^^
```

### Why are the changes needed?
To improve error messages for better usability.

### Does this PR introduce _any_ user-facing change?
If error messages are considered as user-facing changes, then yes. It improves the messages as above.

### How was this patch tested?
It's tested by newly added unit tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala [hunk 1]
```diff
@@ -221,6 +221,11 @@ object QueryParsingErrors {
     new ParseException(s"DataType $dataType is not supported.", ctx)
   }
 
+  def charTypeMissingLengthError(dataType: String, ctx: PrimitiveDataTypeContext): Throwable = {
+    new ParseException(s"DataType $dataType requires a length parameter, for example " +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala:225]: shall we create an error class for it?

Comment (anchovYu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala]: sure!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ErrorParserSuite.scala [hunk 2]
```diff
@@ -255,4 +255,17 @@ class ErrorParserSuite extends AnalysisTest {
         |SELECT b
       """.stripMargin, 2, 9, 10, msg + " test-table")
   }
+
+  test("datatype not supported") {
+    // general bad types
+    intercept("SELECT cast(1 as badtype)", 1, 17, 17, "DataType badtype is not supported.")
+
+    // special handling on char and varchar
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ErrorParserSuite.scala:263]: Can we put the tests in `QueryParsingErrorsSuite`? Which has the infra to check error class

Comment (anchovYu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ErrorParserSuite.scala:263]: This class also has the `intercept` that can check the error class (see the below `PARSE_CHAR_MISSING_LENGTH`). The current tests check whether the error class is `PARSE_CHAR_MISSING_LENGTH`.
Wonder what's the difference between this `ErrorParserSuite` and `QueryParsingErrorsSuite`? Seems they all testing parsing exceptions, but this `ErrorParserSuite` also checks the ANTLR errors.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.3!

**Pull request closed without merging**